core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java

```diff
@@ -0,0 +1,213 @@
+        Types.NestedField.required(1, Column.PARTITION_DATA.name(), partitionType),
+        Types.NestedField.required(2, Column.SPEC_ID.name(), Types.IntegerType.get()),
+        Types.NestedField.required(3, Column.DATA_RECORD_COUNT.name(), Types.LongType.get()),
+        Types.NestedField.required(4, Column.DATA_FILE_COUNT.name(), Types.IntegerType.get()),
+        Types.NestedField.required(5, Column.DATA_FILE_SIZE_IN_BYTES.name(), Types.LongType.get()),
+        Types.NestedField.optional(
+            6, Column.POSITION_DELETE_RECORD_COUNT.name(), Types.LongType.get()),
+        Types.NestedField.optional(
+            7, Column.POSITION_DELETE_FILE_COUNT.name(), Types.IntegerType.get()),
+        Types.NestedField.optional(
+            8, Column.EQUALITY_DELETE_RECORD_COUNT.name(), Types.LongType.get()),
+        Types.NestedField.optional(
+            9, Column.EQUALITY_DELETE_FILE_COUNT.name(), Types.IntegerType.get()),
+        Types.NestedField.optional(10, Column.TOTAL_RECORD_COUNT.name(), Types.LongType.get()),
+        Types.NestedField.optional(11, Column.LAST_UPDATED_AT.name(), Types.LongType.get()),
+        Types.NestedField.optional(
+            12, Column.LAST_UPDATED_SNAPSHOT_ID.name(), Types.LongType.get()));
+  }
+
+  public static CloseableIterable<Record> fromManifest(
+      Table table, ManifestFile manifest, Schema recordSchema) {
+    CloseableIterable<? extends ManifestEntry<? extends ContentFile<?>>> entries =
+        CloseableIterable.transform(
+            ManifestFiles.open(manifest, table.io(), table.specs())
+                .select(scanColumns(manifest.content())) // don't select stats columns
+                .liveEntries(),
+            t ->
+                (ManifestEntry<? extends ContentFile<?>>)
+                    // defensive copy of manifest entry without stats columns
+                    t.copyWithoutStats());
+
+    return CloseableIterable.transform(
+        entries, entry -> fromManifestEntry(entry, table, recordSchema));
+  }
+
+  public static void updateRecord(Record toUpdate, Record fromRecord) {
+    toUpdate.set(
+        Column.SPEC_ID.ordinal(),
+        Math.max(
+            (int) toUpdate.get(Column.SPEC_ID.ordinal()),
+            (int) fromRecord.get(Column.SPEC_ID.ordinal())));
+    incrementLong(toUpdate, fromRecord, Column.DATA_RECORD_COUNT);
+    incrementInt(toUpdate, fromRecord, Column.DATA_FILE_COUNT);
+    incrementLong(toUpdate, fromRecord, Column.DATA_FILE_SIZE_IN_BYTES);
+    checkAndIncrementLong(toUpdate, fromRecord, Column.POSITION_DELETE_RECORD_COUNT);
+    checkAndIncrementInt(toUpdate, fromRecord, Column.POSITION_DELETE_FILE_COUNT);
+    checkAndIncrementLong(toUpdate, fromRecord, Column.EQUALITY_DELETE_RECORD_COUNT);
+    checkAndIncrementInt(toUpdate, fromRecord, Column.EQUALITY_DELETE_FILE_COUNT);
+    checkAndIncrementLong(toUpdate, fromRecord, Column.TOTAL_RECORD_COUNT);
+    if (toUpdate.get(Column.LAST_UPDATED_AT.ordinal()) != null
```

Comment:
Same question as below, shouldn't we set toUpdate to the value of fromRecord if toUpdate is null but fromRecord is not?